File: lc/q3-longest-substring-without-repeating-char/Solution.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0)
            return 0;
        Map<Character, Integer> map = new HashMap<>();
        int max = 0;

        for (int i = 0, j = 0; i < s.length(); i++) {
            if (map.containsKey(s.charAt(i))) {
                j = Math.max(j, map.get(s.charAt(i)) + 1);
            }

            map.put(s.charAt(i), i);
            max = Math.max(max, i - j + 1);
        }

        return max;
    }

    public static void main(String[] args) {
        Solution sol = new Solution();
        System.out.println(sol.lengthOfLongestSubstring("abcabcbb"));
        System.out.println(sol.lengthOfLongestSubstring("bbbbb"));
        System.out.println(sol.lengthOfLongestSubstring("pwwkew"));
        System.out.println(sol.lengthOfLongestSubstring(""));
        System.out.println(sol.lengthOfLongestSubstring("abba"));
    }
}

File: lc/q3-longest-substring-without-repeating-char/solution.py
from collections import defaultdict


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if len(s) == 0:
            return 0

        _map = defaultdict(int)
        i = 0
        _max = 0
        left_p = 0

        while i < len(s):
            char = s[i]

            if char in _map:
                # update left_p index
                left_p = max(left_p, _map[char] + 1)

            _map[char] = i
            _max = max(_max, i - left_p + 1)
            i += 1

        return _max


if __name__ == "__main__":
    s = Solution()

    # Example 1
    assert s.lengthOfLongestSubstring("abcabcbb") == 3

    # Example 2
    assert s.lengthOfLongestSubstring("bbbbb") == 1

    # Example 3
    assert s.lengthOfLongestSubstring("pwwkew") == 3

    # Example 4
    assert s.lengthOfLongestSubstring("") == 0

    # Example 5
    assert s.lengthOfLongestSubstring("abba") == 2

    print("All passed")
